### PR TITLE
refactor(moduletypes): introduce module type constants for plugins

### PR DESCRIPTION
**Pull request by TheLarkInn**
## Summary 

This PR introduces a new abstraction to the usages of module types passed to parser hooks. 

## Details 

Currently today we are wasting a lot of memory using hundreds/thousands of the same string literals which can add up overtime. Also in the codebase there is no proper documentation for what each module type strings actually represent. 

This abstraction (and implementation in the DefinePlugin and JSONModulesPlugin) aim to improve understanding of these module types, provide improved type safety, and reduces overusage of memory for string literals. 

This PR also does the following: 
* Abstract plugin name in DefinePlugin and JSONModulesPlugin
* Abstract module types usages 
* hoists any RegExp literals and gives them somewhat useful names.

### CODE REVIEW

File: lib/DefinePlugin.js [hunk 1]
```diff
@@ -249,8 +255,12 @@ const toCacheVersion = code => {
...
 };
 
-const VALUE_DEP_PREFIX = "webpack/DefinePlugin ";
-const VALUE_DEP_MAIN = "webpack/DefinePlugin_hash";
+const PLUGIN_NAME = "DefinePlugin";
+const VALUE_DEP_PREFIX = `webpack/${PLUGIN_NAME} `;
+const VALUE_DEP_MAIN = `webpack/${PLUGIN_NAME}_hash`;
+const TYPEOF_OPERATOR_REGEXP = /^typeof\s+/;
+const WEBPACK_REQUIRE_FUNCTION_REGEXP = /__webpack_require__\s*(!?\.)/;
+const WEBPACK_REQUIRE_IDENTIFIER_REGEXP = /__webpack_require__/;
```
Comment (alexander-akait) [inline lib/DefinePlugin.js:263]: We can reuse it from https://github.com/webpack/webpack/blob/main/lib/RuntimeGlobals.js#L11

Comment (TheLarkInn) [reply lib/DefinePlugin.js:263]: Yup, that is going to be my next goal. We can leave for now, but thats my next refactoring.

Comment (TheLarkInn) [reply lib/DefinePlugin.js:263]: __webpack_require__ token strings literals are _everywhere_

Comment (TheLarkInn) [reply lib/DefinePlugin.js:263]: Created an issue from your comment and will resolve this.

**Review by alexander-akait — APPROVED** [review]
Great, let's do it later, when we will replace all `webpack_require` tokens

**Pull request merged**